sam_view.c

```diff
@@ -367,7 +371,7 @@ int main_samview(int argc, char *argv[])
         strcat(out_mode, tmp);
         strcat(out_un_mode, tmp);
     }
-    if (argc == optind && isatty(STDIN_FILENO)) return usage(stdout, EXIT_SUCCESS, is_long_help); // potential memory leak...
+    if (argc == optind && isatty(STDIN_FILENO)) return usage(stderr, EXIT_FAILURE, is_long_help); // potential memory leak...
```

Comment:
Now that this is an error message, it makes no sense for it to depend on whether stdin is a tty or not.

Previously this line was there so that you could say `stuff | samtools view -o foo.bam` and it would copy stdin to _foo.bam_ but if you said `samtools view` you would get a usage display. If you really wanted to copy from stuff you typed on the terminal, you could say `samtools view -o foo.bam -`.

Now that the plain `samtools view` usage is handled separately, either this line should disappear or the `if` condition should be as is and the error message should say what's really going on:

> Refusing to sit here reading from standard input and looking like I'm in an infinite loop. If you really want to type my standard input on the terminal, add `-` as a filename

Reply:
Good point about needing `isatty` in the new bit at the top: the cases that are being differentiated include `samtools view` to show the usage and `stuff-outputting-BAM | samtools view` to show SAM records.

It's unsurprising that samtools's argument handling contains error handling bugs: it is old code written by people who did not consider detecting user error to be a priority. For similar reasons, Unix filter conventions and `-` handling are uneven.

Samtools `view` has better argument handling than some of the other samtools commands and is especially heavily used in the middle of pipelines (compared to other subcommands). So IMHO avoiding breaking existing argument handling behaviours **should** hold up this PR :smile:

Reply:
Oh I agree that breaking things should hold it up!  We shouldn't go backwards.

I just meant that not fixing other existing bugs isn't necessarily a problem.  So for example `cat foo.bam | samtools view -r` doesn't fail.  It was true before, and it's still true now.  Ideally we'd fix all these in one step, but if split over subsequent PRs I could live with that too.